src/stream/src/executor/source/fetch_executor.rs

```diff
@@ -0,0 +1,367 @@
+// limitations under the License.
+
+use std::fmt::{Debug, Formatter};
+use std::ops::Bound;
+use std::sync::Arc;
+
+use either::Either;
+use futures::pin_mut;
+use futures::stream::{self, StreamExt};
+use futures_async_stream::try_stream;
+use risingwave_common::catalog::{ColumnId, Schema, TableId};
+use risingwave_common::hash::VnodeBitmapExt;
+use risingwave_common::row::{OwnedRow, Row};
+use risingwave_common::types::{ScalarRef, ScalarRefImpl};
+use risingwave_connector::source::filesystem::FsSplit;
+use risingwave_connector::source::{
+    BoxSourceWithStateStream, SourceContext, SourceCtrlOpts, SplitImpl, SplitMetaData,
+    StreamChunkWithState,
+};
+use risingwave_connector::ConnectorParams;
+use risingwave_source::source_desc::SourceDesc;
+use risingwave_storage::store::PrefetchOptions;
+use risingwave_storage::StateStore;
+
+use crate::executor::stream_reader::StreamReaderWithPause;
+use crate::executor::{
+    expect_first_barrier, ActorContextRef, BoxedExecutor, BoxedMessageStream, Executor, Message,
+    Mutation, PkIndices, PkIndicesRef, SourceStateTableHandler, StreamExecutorError,
+    StreamExecutorResult, StreamSourceCore,
+};
+
+const SPLIT_BATCH_SIZE: usize = 1000;
+
+type SplitBatch = Option<Vec<SplitImpl>>;
+
+pub struct FsFetchExecutor<S: StateStore> {
+    actor_ctx: ActorContextRef,
+
+    identity: String,
+
+    schema: Schema,
+
+    pk_indices: PkIndices,
+
+    /// Streaming source for external
+    stream_source_core: Option<StreamSourceCore<S>>,
+
+    /// Upstream list executor.
+    upstream: Option<BoxedExecutor>,
+
+    // control options for connector level
+    source_ctrl_opts: SourceCtrlOpts,
+
+    // config for the connector node
+    connector_params: ConnectorParams,
+}
+
+impl<S: StateStore> FsFetchExecutor<S> {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        actor_ctx: ActorContextRef,
+        schema: Schema,
+        pk_indices: PkIndices,
+        stream_source_core: StreamSourceCore<S>,
+        executor_id: u64,
+        upstream: BoxedExecutor,
+        source_ctrl_opts: SourceCtrlOpts,
+        connector_params: ConnectorParams,
+    ) -> Self {
+        Self {
+            actor_ctx,
+            identity: format!("FsFetchExecutor {:X}", executor_id),
+            schema,
+            pk_indices,
+            stream_source_core: Some(stream_source_core),
+            upstream: Some(upstream),
+            source_ctrl_opts,
+            connector_params,
+        }
+    }
+
+    async fn try_replace_with_new_batch_reader<const BIASED: bool>(
+        splits_on_fetch: &mut usize,
+        state_store_handler: &SourceStateTableHandler<S>,
+        column_ids: Vec<ColumnId>,
+        source_ctx: SourceContext,
+        source_desc: &SourceDesc,
+        stream: &mut StreamReaderWithPause<BIASED, StreamChunkWithState>,
+    ) -> StreamExecutorResult<()> {
+        if *splits_on_fetch > 0 {
+            return Ok(());
+        }
```

Comment:
I dont think it is a good design because the stream may not be replaced after we call the func. you can do the check outside.